tools/wptrunner/wptrunner/browsers/sauce.py

```diff
@@ -127,15 +127,33 @@ def get_tar(url, dest):
 class SauceConnect():
 
     def __init__(self, **kwargs):
-        self.config = kwargs["config"]
         self.sauce_user = kwargs["sauce_user"]
         self.sauce_key = kwargs["sauce_key"]
         self.sauce_tunnel_id = kwargs["sauce_tunnel_id"]
         self.sauce_connect_binary = kwargs.get("sauce_connect_binary")
         self.sc_process = None
         self.temp_dir = None
 
-    def __enter__(self, options):
+        # Because this class implements the context manager protocol, it is
+        # possible for instances to be provided to the `with` statement
+        # directly. The `use` method is defined so that data which is not
+        # available during object initialization can be provided prior to this
+        # moment. The `use` method must be invoked in preparation for the
+        # context manager protocol, but this additional constraint is not
+        # itself part of the protocol. A dedicated flag is maintained in order
+        # to provide a more descriptive error to callers who fail to honor this
+        # constraint.
+        self.use_invoked = False
+
+    def use(self, env_options, env_config):
+        self.use_invoked = True
```

Comment:
Presumably this flag should be unset on `__exit__`. But rather than having a dedicated flag, why not use the fact that `env_config` is initially unset?

Reply:
The thought occurred to me, but I decided an explicit flag would be easier to understand and harder to accidentally break. This is pretty subjective, though, so I'll re-implement as you suggest. Would you like me to unset `self.env_config` in `__exit__`?

Reply:
I'm happy with either.